File: library/src/main/java/com/xxxifan/devbox/library/util/Fragments.java
/*
 * Copyright  (c) 2016 xxxifan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.xxxifan.devbox.library.util;

import android.annotation.SuppressLint;
import android.support.annotation.AnimRes;
import android.support.annotation.CheckResult;
import android.support.annotation.IdRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentTransaction;
import android.view.View;

import com.orhanobut.logger.Logger;
import com.xxxifan.devbox.library.base.BaseFragment;

import java.lang.ref.WeakReference;
import java.util.List;

/**
 * Created by xifan on 6/7/16.
 */
public class Fragments {
    public static final String TAG = "Fragments";
    public static final String KEY_RESTORE = "restore";
    public static final String KEY_RESTORE_VIEWPAGER = "restore_viewpager";

    private Fragments() {
    }

    /**
     * checkout with FRAGMENT_CONTAINER(which is defined in BaseActivity, is R.id.fragment_container
     * it will use BaseFragment.getSimpleName() as tag, or SimpleClassName if fallback.
     */
    @CheckResult
    public static SingleOperator checkout(FragmentActivity activity, Fragment fragment) {
        return new SingleOperator(activity, fragment);
    }

    /**
     * checkout with specified tag
     */
    @CheckResult
    public static SingleOperator checkout(FragmentActivity activity, Fragment fragment, String tag) {
        return new SingleOperator(activity, fragment, tag);
    }

    /**
     * checkout previously fragment by tag
     */
    @CheckResult
    public static SingleOperator checkout(FragmentActivity activity, String tag) {
        return new SingleOperator(activity, tag);
    }

    /**
     * add multi fragments
     */
    @CheckResult
    public static MultiOperator add(FragmentActivity activity, Fragment... fragments) {
        if (fragments == null) {
            throw new IllegalArgumentException("Can't accept null fragments");
        }
        return new MultiOperator(activity, fragments);
    }

    /**
     * get current visible fragment on container
     */
    public static Fragment getCurrentFragment(FragmentActivity activity, int containerId) {
        return activity.getSupportFragmentManager().findFragmentById(containerId);
    }

    public static Fragment getFragment(FragmentActivity activity, String tag) {
        return activity.getSupportFragmentManager().findFragmentByTag(tag);
    }

    public static List<Fragment> getFragmentList(FragmentActivity activity) {
        return activity.getSupportFragmentManager().getFragments();
    }

    private static String getTag(Fragment fragment) {
        return Strings.isEmpty(fragment.getTag())
                ? (fragment instanceof BaseFragment ? ((BaseFragment) fragment).getSimpleName() : fragment
                .getClass().getName()) : fragment.getTag();
    }

    public static class SingleOperator {
        private WeakReference<FragmentActivity> activityRef;
        private Fragment fragment;
        private String tag;
        private FragmentTransaction transaction;

        private boolean addToBackStack;
        private boolean fade;
        private boolean removeLast;
        private boolean hideLast = true;

        private SingleOperator(FragmentActivity activity, Fragment fragment) {
            this(activity, fragment, getTag(fragment));
        }

        @SuppressLint("CommitTransaction")
        private SingleOperator(FragmentActivity activity, Fragment fragment, String tag) {
            this.activityRef = new WeakReference<>(activity);
            this.fragment = fragment;
            this.tag = tag;
            this.transaction = activity.getSupportFragmentManager().beginTransaction();
        }

        @SuppressLint("CommitTransaction")
        private SingleOperator(FragmentActivity activity, String tag) {
            this.activityRef = new WeakReference<>(activity);
            this.tag = tag;
            this.transaction = activity.getSupportFragmentManager().beginTransaction();

            // retrieve correct fragment
            List<Fragment> fragments = getFragmentList(activity);
            for (Fragment tagFragment : fragments) {
                if (Strings.equals(tagFragment.getTag(), tag)) {
                    this.fragment = tagFragment;
                    break;
                }
            }
        }

        public SingleOperator addSharedElement(View sharedElement, String name) {
            transaction.addSharedElement(sharedElement, name);
            return this;
        }

        public SingleOperator setCustomAnimator(@AnimRes int enter, @AnimRes int exit) {
            transaction.setCustomAnimations(enter, exit);
            return this;
        }

        public SingleOperator setCustomAnimator(@AnimRes int enter, @AnimRes int exit, @AnimRes int popEnter, @AnimRes int popExit) {
            transaction.setCustomAnimations(enter, exit, popEnter, popExit);
            return this;
        }

        public SingleOperator addToBackStack(boolean add) {
            this.addToBackStack = add;
            return this;
        }

        /**
         * display fade transition
         */
        public SingleOperator fade() {
            this.fade = true;
            return this;
        }

        /**
         * hideLast last fragment, default is true.
         * if you want last to remove, see {@link #removeLast(boolean)}
         */
        public SingleOperator hideLast(boolean hideLast) {
            this.hideLast = hideLast;
            return this;
        }

        /**
         * remove last fragment while checkout.
         */
        public SingleOperator removeLast(boolean remove) {
            this.removeLast = remove;
            return this;
        }

        public void into(@IdRes int containerId) {
            if (fragment == null) {
                Logger.t(TAG).e("fragment is null, will not do anything");
                finish();
                return;
            }

            // hide or remove last fragment
            if (hideLast || removeLast) {
                List<Fragment> fragments = getFragmentList(activityRef.get());
                if (fragments != null) {
                    for (Fragment oldFragment : fragments) {
                        if (oldFragment == null) {
                            continue;
                        }

                        if (oldFragment.getId() == containerId) {
                            if (Strings.equals(oldFragment.getTag(), tag)) {
                                fragment = oldFragment;
                            } else if (oldFragment.isVisible()) {
                                oldFragment.setUserVisibleHint(false);
                                transaction.hide(oldFragment);
                                if (removeLast) {
                                    Logger.d("last fragment has been totally removed");
                                    transaction.remove(oldFragment);
                                }
                            }
                        }
                    }
                }
            }

            boolean canAddBackStack = transaction.isAddToBackStackAllowed() && !transaction.isEmpty();

            if (fade) {
                // noinspection WrongConstant
                transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
            }

            if (!fragment.isAdded()) {
                transaction.add(containerId, fragment, tag);
            }

            transaction.show(fragment);

            if (addToBackStack) {
                if (canAddBackStack) {
                    transaction.addToBackStack(tag);
                } else {
                    Logger.t(TAG)
                            .w("addToBackStack called, but this is not permitted");
                }
            }

            finish();
        }

        private void finish() {
            transaction.commitAllowingStateLoss();
            transaction = null;
            fragment = null;
            activityRef.clear();
        }
    }

    // TODO: 6/10/16 MultiOperator is not used that much, so I only give it basic into function here.
    public static class MultiOperator {
        private Fragment[] fragments;
        private WeakReference<FragmentActivity> activityRef;

        @SuppressLint("CommitTransaction")
        public MultiOperator(FragmentActivity activity, Fragment[] fragments) {
            this.fragments = fragments;
            activityRef = new WeakReference<>(activity);
        }

        public void into(int... ids) {
            if (ids.length != fragments.length) {
                throw new IllegalArgumentException("The length of ids and fragments is not equal.");
            }

            FragmentTransaction transaction = activityRef.get()
                    .getSupportFragmentManager()
                    .beginTransaction();
            String tag;
            for (int i = 0, s = ids.length; i < s; i++) {
                tag = getTag(fragments[i]);
                transaction.replace(ids[i], fragments[i], tag);
            }
            transaction.commitAllowingStateLoss();

            activityRef.clear();
            fragments = null;
        }
    }
}

File: library/src/main/java/com/xxxifan/devbox/core/util/Fragments.kt
package com.xxxifan.devbox.core.util

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Parcelable
import android.view.View
import androidx.annotation.AnimRes
import androidx.annotation.CheckResult
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentTransaction
import androidx.viewpager.widget.ViewPager
import com.xxxifan.devbox.core.base.BasePresenter
import com.xxxifan.devbox.core.base.BaseView

/**
 * Created by xifan on 6/7/16.
 */
@SuppressLint("RestrictedApi")
object Fragments {
  const val TAG = "Fragments"
  private const val KEY_RESTORE = "restore"
  private const val KEY_RESTORE_VIEWPAGER = "restore_viewpager"

  /**
   * checkout fragment into container layout.
   * it will use BaseFragment.getSimpleName() as tag, or Class.name as fallback. For more see
   * [Operator]
   */
  @JvmOverloads
  @CheckResult
  fun checkout(
    activity: FragmentActivity, fragment: Fragment? = null,
    tag: String? = null
  ): Operator<FragmentActivity> {
    return Operator(activity, fragment, tag)
  }

  /**
   * [checkout] for fragment host
   */
  @JvmOverloads
  @CheckResult
  fun checkout(
    hostFragment: Fragment,
    childFragment: Fragment? = null, tag: String? = null
  ): Operator<Fragment> {
    return Operator(hostFragment, childFragment, tag)
  }

  fun hide(hostActivity: FragmentActivity, fragments: Array<Fragment>) {
    val transac = hostActivity.supportFragmentManager.beginTransaction()
    fragments.forEach {
      transac.hide(it)
      it.userVisibleHint = false
    }
    transac.commitAllowingStateLoss()
  }

  fun hide(hostFragment: Fragment, fragments: Array<Fragment>) {
    val trans = hostFragment.childFragmentManager.beginTransaction()
    fragments.forEach {
      trans.hide(it)
      it.userVisibleHint = false
    }
    trans.commitAllowingStateLoss()
  }

  /**
   * get current visible fragment on container
   */
  fun getCurrentFragment(activity: FragmentActivity, containerId: Int): Fragment? {
    return activity.supportFragmentManager.findFragmentById(containerId)
  }

  fun getFragment(activity: FragmentActivity, tag: String): Fragment? {
    return activity.supportFragmentManager.findFragmentByTag(tag)
  }

  fun getFragmentList(activity: FragmentActivity): List<Fragment> {
    return activity.supportFragmentManager.fragments
  }

  fun getChildFragmentList(fragment: Fragment): List<Fragment> {
    return fragment.childFragmentManager.fragments
  }

  fun onSaveInstanceState(fragment: Fragment, outState: Bundle) {
    outState.putBoolean(KEY_RESTORE, fragment.isVisible)
    outState.putBoolean(KEY_RESTORE_VIEWPAGER, fragment.view?.parent is ViewPager)
  }

  fun restoreFragmentState(fragment: Fragment, outState: Bundle?) {
    if (outState == null) return
    if (outState.getBoolean(KEY_RESTORE_VIEWPAGER, false)) return
    if (!fragment.isAdded) return
    val transaction = fragment.parentFragmentManager.beginTransaction()

    if (outState.getBoolean(KEY_RESTORE, false)) {
      transaction.show(fragment)
    } else {
      transaction.hide(fragment)
    }
    transaction.commitAllowingStateLoss()
  }

  private fun getTag(fragment: Fragment): String {
    return if (fragment.tag.isNullOrBlank()) fragment.javaClass.name else fragment.tag ?: ""
  }

  @SuppressLint("CommitTransaction")
  class Operator<HostType>
  internal constructor(
    host: HostType,
    private var fragment: Fragment? = null, private var tag: String? = null
  ) {
    private var fragments: List<Fragment?>
    private var transaction: FragmentTransaction? = null
    private var hostTag: String

    // config field
    private var presenter: BasePresenter<Any>? = null
    private var addToBackStack = false
    private var backStackTag: String? = null
    private var fade = false
    private var retainInstance = false

    init {
      val hostActivity = host as? FragmentActivity
      val hostFragment = host as? Fragment
      when {
        hostActivity != null -> {
          transaction = hostActivity.supportFragmentManager.beginTransaction()
          fragments = getFragmentList(hostActivity)
          hostTag = hostActivity.localClassName
        }
        hostFragment != null -> {
          transaction = hostFragment.childFragmentManager.beginTransaction()
          fragments = getChildFragmentList(hostFragment)
          hostTag = getTag(hostFragment)
        }
        else -> throw RuntimeException(
          "host must be androidx.fragment.app.FragmentActivity or androidx.fragment.app.Fragment"
        )
      }

//      transaction!!.setReorderingAllowed(true)

      if (tag?.isNotBlank() == true && fragment == null) { // retrieve correct fragment
        fragment = fragments.firstOrNull { it?.tag == tag }
      } else if (tag.isNullOrBlank() && fragment != null) { // retrieve correct tag
        tag = getTag(fragment!!)
      }
    }

    fun <T : BasePresenter<*>?> bindPresenter(presenter: T?): Operator<HostType> {
      presenter?.let { this.presenter = it as BasePresenter<Any> }
      return this
    }

    /**
     * setArguments to target fragment.
     */
    fun data(data: Bundle?): Operator<HostType> {
      if (fragment?.isStateSaved == true) return this // make sure fragment state.
      data ?: return this

      fragment?.arguments = data
      return this
    }

    /**
     * simple string bundle as argument
     */
    fun data(key: String, value: String?): Operator<HostType> {
      if (fragment?.isStateSaved == true) return this // make sure fragment state.

      val bundle = fragment?.arguments ?: Bundle()
      bundle.putString(key, value)
      fragment?.arguments = bundle
      return this
    }

    fun data(key: String, value: Boolean): Operator<HostType> {
      if (fragment?.isStateSaved == true) return this // make sure fragment state.

      val bundle = fragment?.arguments ?: Bundle()
      bundle.putBoolean(key, value)
      fragment?.arguments = bundle
      return this
    }

    fun data(key: String, value: Parcelable?): Operator<HostType> {
      if (fragment?.isStateSaved == true) return this // make sure fragment state.

      val bundle = fragment?.arguments ?: Bundle()
      bundle.putParcelable(key, value)
      fragment?.arguments = bundle
      return this
    }

    /**
     * see [FragmentTransaction.addSharedElement]
     */
    fun addSharedElement(sharedElement: View, name: String): Operator<HostType> {
      transaction!!.addSharedElement(sharedElement, name)
      return this
    }

    /**
     * see [FragmentTransaction.setCustomAnimations]
     */
    fun setCustomAnimator(@AnimRes enter: Int, @AnimRes exit: Int): Operator<HostType> {
      transaction!!.setCustomAnimations(enter, exit)
      return this
    }

    /**
     * see [FragmentTransaction.setCustomAnimations]
     */
    fun setCustomAnimator(
      @AnimRes enter: Int, @AnimRes exit: Int,
      @AnimRes popEnter: Int, @AnimRes popExit: Int
    ): Operator<HostType> {
      transaction!!.setCustomAnimations(enter, exit, popEnter, popExit)
      return this
    }

    /**
     * Fragments use transaction optimization for better performance, if you face issues please
     * disable it.
     */
//    fun disableOptimize(): Operator<HostType> {
//      transaction!!.setAllowOptimization(false)
//      return this
//    }

    fun addToBackStack(tag: String? = null): Operator<HostType> {
      this.addToBackStack = true
      this.backStackTag = tag
      return this
    }

    /**
     * make fragment call [Fragment.setRetainInstance(true)]
     */
    fun retainInstance(): Operator<HostType> {
      this.retainInstance = true
      return this
    }

    /**
     * display fade transition
     */
    fun fade(): Operator<HostType> {
      this.fade = true
      return this
    }

    /**
     * @return success or not
     */
    fun into(@IdRes containerId: Int): Boolean {
      val fragment = this.fragment
      if (fragment == null) {
        commit()
        return false
      }

      // use local val for easier access
      val fragments = this.fragments
      val transaction = this.transaction!!

      // hide other fragment
      fragments.asSequence()
        .filter { it != null && it.id == containerId && it.isAdded }
        .forEach {
          if (it!!.tag != tag && it.isVisible) {
            transaction.hide(it)
            it.userVisibleHint = false
          }
        }

      val canAddBackStack = transaction.isAddToBackStackAllowed && !transaction.isEmpty
      if (addToBackStack && canAddBackStack) {
        transaction.addToBackStack(backStackTag)
      }
      if (fade) {
        transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
      }
      if (!fragment.isAdded) {
        transaction.add(containerId, fragment, tag)
      }
      if (retainInstance) {
        fragment.retainInstance = true
      }
      presenter?.view = (fragment as? BaseView)

      fragment.userVisibleHint = true
      transaction.show(fragment)

      commit()
      return true
    }

    private fun commit() {
      transaction!!.commitAllowingStateLoss()
      transaction = null
      fragment = null
      presenter = null
    }
  }
}

fun Fragment.hide() {
  if (!isAdded) return
  val trans = parentFragmentManager.beginTransaction()
  trans.hide(this)
  trans.commit()
}
fun Fragment.remove() {
  if (!isAdded) return
  val trans = parentFragmentManager.beginTransaction()
  trans.remove(this)
  trans.commit()
}